spec/lib/pedicel/validator_spec.rb

```diff
@@ -0,0 +1,513 @@
+require 'pedicel/validator'
+require 'pedicel-pay'
+
+require 'json'
+require 'digest'
+
+valid_ec_public_key = 'MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE0wzW/i0nT3XOeo6srobJRGnUlmqGTFahuHEOw4M9nxUlTBaQUNc8HlN/z1HbepGXTZWDSJB2deCGfhsrOdVryQ=='
+
+describe Pedicel::Validator do
+  let (:token) do
+    backend = PedicelPay::Backend.generate
+    client = backend.generate_client
+
+    token = PedicelPay::Token.new.sample
+    backend.encrypt_and_sign(token, recipient: client)
+
+    token
+  end
+
+  describe '.valid_token?' do
+    it 'relies on validate_token' do
+      expect(Pedicel::Validator).to receive(:validate_token).with(nil).and_return(true)
+
+      expect(Pedicel::Validator.valid_token?(nil)).to eq(true)
+    end
+
+    it 'relies on validate_token' do
+      expect(Pedicel::Validator).to receive(:validate_token).with(nil).and_raise(Pedicel::Validator::Error, 'boom')
+
+      expect(Pedicel::Validator.valid_token?(nil)).to eq(false)
+    end
+  end
+
+  describe '.validate_token' do
+    context 'can be happy' do
+      subject { lambda { Pedicel::Validator.validate_token(token.to_hash) } }
+
+      it 'does not err on a valid token' do
+        is_expected.to_not raise_error
+      end
+
+      it 'is truthy a valid token' do
+        expect(Pedicel::Validator.validate_token(token.to_hash)).to be_truthy
+      end
+    end
+
+    let (:token_h) { token.to_hash }
+    subject { lambda { Pedicel::Validator.validate_token(token_h) } }
+
+    context 'wrong data' do
+      it 'errs when data is missing' do
+        token_h.delete('data')
+        is_expected.to raise_error(Pedicel::Validator::TokenFormatError, /data:/)
+      end
+
+      it 'errs when data is missing' do
+        token_h.delete('data')
+        is_expected.to raise_error(Pedicel::Validator::TokenFormatError, /data:.*missing/)
+      end
```

Comment:
Duplicate test cases? Lines 51-59.

Reply:
🤦‍♂️
Will fix. By removing the sloppier test.